master/internal/core_trial.go

```diff
@@ -40,6 +45,123 @@ func echoCanGetTrial(c echo.Context, m *Master, trialID string) error {
+	if len(in) == 0 {
+		return nil, echo.NewHTTPError(http.StatusBadRequest,
+			"expected at least one trial ID in the query param 'trialIds'")
+	}
+
+	var trialIDs []int
+	for _, id := range strings.Split(in, ",") {
+		if err := echoCanGetTrial(c, m, id); err != nil {
+			return nil, err
+		}
+
+		trialID, err := strconv.Atoi(id)
+		if err != nil {
+			return nil, echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf(
+				"could not parse trialIds parameter expected comma separated ints got %s", in))
+		}
+		trialIDs = append(trialIDs, trialID)
+	}
+
+	return trialIDs, nil
+}
+
+//	@Summary	Stream one or more trial's training metrics.
+//	@Tags		Trials
+//	@ID			stream-training-metrics
+//	@Produce	application/x-ndjson
+//	@Param		trialIds	query	string	true	"Comma delimited trial IDs"
+//	@Success	200					{}		string	"A new line delimated JSON stream with the following fields trialId,endTime,metrics,totalBatches,trialRunId,id"
+//	@Router		/trials/metrics/training [get]
+//
+// nolint:lll
+func (m *Master) streamTrainingMetrics(c echo.Context) error {
+	trialIDs, err := m.parseAndRBACTrialIDs(c, c.QueryParam("trialIds"))
+	if err != nil {
+		return err
+	}
+
+	err = streamQueryAsNDJSON(c, `SELECT jsonb_build_object(
+		'trialId', trial_id,
+		'endTime', end_time,
+		'metrics', metrics,
+		'totalBatches', total_batches,
+		'trialRunId', trial_run_id,
+		'id', id
+	) FROM steps WHERE trial_id IN (?) ORDER BY trial_id, trial_run_id, total_batches`,
+		bun.In(trialIDs))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			errors.Wrap(err, "error streaming training metrics"))
+	}
+	return nil
+}
+
+//	@Summary	Stream one or more trial's validation metrics.
+//	@Tags		Trials
+//	@ID			stream-validation-metrics
+//	@Produce	application/x-ndjson
+//	@Param		trialIds	query	string	true	"Comma delimited trial IDs"
+//	@Success	200					{}		string	"A new line delimated JSON stream with the following fields trialId,endTime,metrics,totalBatches,trialRunId,id"
+//	@Router		/trials/metrics/validation [get]
+//
+// nolint:lll
+func (m *Master) streamValidationMetrics(c echo.Context) error {
+	trialIDs, err := m.parseAndRBACTrialIDs(c, c.QueryParam("trialIds"))
+	if err != nil {
+		return err
+	}
+
+	err = streamQueryAsNDJSON(c, `SELECT jsonb_build_object(
+		'trialId', trial_id,
+		'endTime', end_time,
+		'metrics', metrics,
+		'totalBatches', total_batches,
+		'trialRunId', trial_run_id,
+		'id', id
+	) FROM validations WHERE trial_id IN (?) ORDER BY trial_id, trial_run_id, total_batches`,
+		bun.In(trialIDs))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			errors.Wrap(err, "error streaming training metrics"))
+	}
+	return nil
+}
+
+func streamQueryAsNDJSON(c echo.Context, query string, args ...any) error {
+	rows, err := db.Bun().QueryContext(c.Request().Context(), query, args...)
+	if err != nil {
+		return errors.Wrap(err, "error running query to stream")
```

Comment:
strictly as nice to have, no worries if this is complex.

when I run this, and interrupt it midway, master logs an error
```
ERRO[2023-03-23T15:42:49-07:00] code=500, message=error streaming training metrics: error running query to stream: timeout: timeout: read tcp [::1]:58971->[::1]:5432: i/o timeout
```
I suspect this will happen every so often, and it's completely harmless case, but people reading the logs may be alarmed by it. it may be nice to catch these, and report as DEBUG or INFO level?